### ISSUE TITLE
Allow the tool to be used in either a Python 2 or Python 3 environment

### ISSUE THREAD

**Issue by konklone** [dhs-ncats/pshtt#4]
This means replacing the use of `sslyze` with something that can work in Python 3, or separating the `sslyze` component out in some way (by having the user rely on Docker or pyenv or something). 

For what the tool currently does (analysis of HTTP behavior and HTTPS failure behavior), I think `sslyze` is wholly replace-able. `sslyze`, or something like it, would be more necessary for analyzing other TLS qualities, like use of ciphers or protocol versions.

Replacing `sslyze` would mean digging into `requests`' API more and analyzing certificate validation errors. This would also have the benefit of speeding up use of the tool in development considerably, because these would be included in the transparent `requests` caching added in #3.

However, we do want to still allow the tool to be used in a Python 2 environment, if at all possible. I believe that's doable.

**Comment by garrettr** [comment]
> This means replacing the use of sslyze with something that can work in Python 3

In an ideal world, this issue would be resolved by making sslyze compatible with Python 3. This would enable us to make pshtt compatible with Python 2.x and 3.x, while maintaining the potential to support future types of scans which are only possible via sslyze.

This is not impossible: the biggest hurdle appears to be porting nassl, the Python C extension that implements the OpenSSL bindings required by sslyze, to be compatible with Python 3. An external contributor has volunteered to do this, and appears to be making progress on their [branch](https://github.com/siteblindado/nassl/tree/python3). I just commented on the relevant upstream issue with the intention of reviving it: https://github.com/nabla-c0d3/sslyze/issues/138.

**Comment by garrettr** [comment]
As stated in the issue description, the easiest way to enable Python 3 support is to remove the dependency on sslyze. To do so, we need to determine if we can at least achieve feature parity between sslyze and another library (preferably `requests`, since it's already being used in pshtt) for the scan results that are currently generated by sslyze.

pshtt derives the following results from sslyze:

* `https_expired_cert`
* `https_bad_chain`
* `https_bad_hostname`

`requests` responds to any SSL errors during a request by raising `SSLError`. This is not as explicit as I would like it to be, but a quick survey of the requests code shows that context-specific information is typically passed to the SSLError constructor, albeit in a cornucopia of types/formats.

Let's see what exceptions are raised by `requests.get` for examples of each of the above types of errors:

```py
import requests

urls = [
    'expired.badssl.com',
    'untrusted-root.badssl.com',
    'wrong.host.badssl.com'
]

for url in urls:
    try:
        requests.get('https://' + url)
    except requests.exceptions.SSLError as e:
        print('{} raised SSLError: {}'.format(e.request.url, e))
```

Output:

```
https://expired.badssl.com/ raised SSLError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:590)
https://untrusted-root.badssl.com/ raised SSLError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:590)
ERROR:requests.packages.urllib3.connection:Certificate did not match expected hostname: wrong.host.badssl.com. Certificate: {'crlDistributionPoints': (u'http://crl.comodoca.com/COMODORSADomainValidationSecureServerCA.crl',), 'subjectAltName': (('DNS', '*.badssl.com'), ('DNS', 'badssl.com')), 'notBefore': u'Jul  7 00:00:00 2016 GMT', 'caIssuers': (u'http://crt.comodoca.com/COMODORSADomainValidationSecureServerCA.crt',), 'OCSP': (u'http://ocsp.comodoca.com',), 'serialNumber': u'4C8E18714B34E75E8DAEFBE8F64C3A82', 'notAfter': 'Sep  5 23:59:59 2017 GMT', 'version': 3L, 'subject': ((('organizationalUnitName', u'Domain Control Validated'),), (('organizationalUnitName', u'PositiveSSL Wildcard'),), (('commonName', u'*.badssl.com'),)), 'issuer': ((('countryName', u'GB'),), (('stateOrProvinceName', u'Greater Manchester'),), (('localityName', u'Salford'),), (('organizationName', u'COMODO CA Limited'),), (('commonName', u'COMODO RSA Domain Validation Secure Server CA'),))}
https://wrong.host.badssl.com/ raised SSLError: hostname 'wrong.host.badssl.com' doesn't match either of '*.badssl.com', 'badssl.com'
```

Unfortunately, this shows that `requests`' error output **does not** sufficiently disambiguate between different types of SSL errors for it to be a suitable replacement for `sslyze`. There is special handling for incorrect hostname errors, but it is impossible to distinguish between an expired cert and a cert that chains to an untrusted root: both raise an SSLError with the insufficiently specific message:

```
[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:590)
```

**Comment by konklone** [comment]
Yes. :( I looked into this a few months back, and came to a similar conclusion. I can't find my notes anymore, but it looked like efforts in the underlying libraries to fix this sort of thing was stymied by `requests` having multiple underlying backends with no standardization to how this sort of information was managed.

It might be possible to go a layer lower below `requests` and use the `ssl` stdlib or the `cryptography` library. I recall looking into this, but getting lost and moving on -- but I think it's probably the most fertile ground to solve this.

**Comment by garrettr** [comment]
> It might be possible to go a layer lower below requests and use the ssl stdlib or the cryptography library. I recall looking into this, but getting lost and moving on -- but I think it's probably the most fertile ground to solve this.

I'm hopeful that the ongoing work of "porting sslyze to Python 3" is the best option here, but agree that switching to ssl stdlib or the cryptography library are also worthwhile possibilities to explore.

**Comment by QuarantineENG** [comment]
FYI tried installing pshtt on CentOs 7


[root@secengcentos7 /]# yum install curl
Loaded plugins: fastestmirror, langpacks
Loading mirror speeds from cached hostfile
 * base: centos.mirror.nac.net
 * extras: linux.cc.lehigh.edu
 * updates: mirrors.mit.edu
Package curl-7.29.0-35.el7.centos.x86_64 already installed and latest version
Nothing to do
[root@secengcentos7 /]# curl "https://bootstrap.pypa.io/get-pip.py" -o "get-pip.py"
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 1558k  100 1558k    0     0  1167k      0  0:00:01  0:00:01 --:--:-- 1168k
[root@secengcentos7 /]# python get-pip.py
Collecting pip
  Downloading pip-9.0.1-py2.py3-none-any.whl (1.3MB)
    100% |████████████████████████████████| 1.3MB 512kB/s 
Collecting wheel
  Downloading wheel-0.29.0-py2.py3-none-any.whl (66kB)
    100% |████████████████████████████████| 71kB 2.7MB/s 
Installing collected packages: pip, wheel
Successfully installed pip-9.0.1 wheel-0.29.0
[root@secengcentos7 /]# pip install pshtt
Collecting pshtt
  Downloading pshtt-0.1.5.tar.gz
Collecting requests>=2.10.0 (from pshtt)
  Downloading requests-2.13.0-py2.py3-none-any.whl (584kB)
    100% |████████████████████████████████| 593kB 1.0MB/s 
Collecting sslyze>=0.13.6 (from pshtt)
  Downloading SSLyze-1.0.0.tar.gz (1.1MB)
    100% |████████████████████████████████| 1.1MB 511kB/s 
Collecting wget>=3.2 (from pshtt)
  Downloading wget-3.2.zip
Collecting docopt (from pshtt)
  Downloading docopt-0.6.2.tar.gz
Collecting requests_cache (from pshtt)
  Downloading requests_cache-0.4.13-py2.py3-none-any.whl
Collecting nassl<0.16.0,>=0.15.0 (from sslyze>=0.13.6->pshtt)
  Downloading nassl-0.15.1.tar.gz (15.2MB)
    100% |████████████████████████████████| 15.2MB 64kB/s 
Collecting typing (from sslyze>=0.13.6->pshtt)
  Downloading typing-3.5.3.0.tar.gz (60kB)
    100% |████████████████████████████████| 61kB 2.4MB/s 
Requirement already satisfied: enum34 in /usr/lib/python2.7/site-packages (from sslyze>=0.13.6->pshtt)
Building wheels for collected packages: pshtt, sslyze, wget, docopt, nassl, typing
  Running setup.py bdist_wheel for pshtt ... done
  Stored in directory: /root/.cache/pip/wheels/a6/03/59/aa30a82beda5aebdbd9a2ecea16169265936db4ab02415c993
  Running setup.py bdist_wheel for sslyze ... done
  Stored in directory: /root/.cache/pip/wheels/ef/73/9e/777d99fb4e3be2e6e7bf462366aefe942a80842d944f85fcc9
  Running setup.py bdist_wheel for wget ... done
  Stored in directory: /root/.cache/pip/wheels/6d/98/29/61ccc41148f871009126c2e844e26f73eeb25e12cca92228a5
  Running setup.py bdist_wheel for docopt ... done
  Stored in directory: /root/.cache/pip/wheels/b2/16/5f/c33a2bb5f2dce71205f8e65cbfd05647d79d441282be31fd82
  Running setup.py bdist_wheel for nassl ... error
  Complete output from command /usr/bin/python -u -c "import setuptools, tokenize;__file__='/tmp/pip-build-Izhzzl/nassl/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" bdist_wheel -d /tmp/tmpm7sp7Ipip-wheel- --python-tag cp27:
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build/lib.linux-x86_64-2.7
  creating build/lib.linux-x86_64-2.7/nassl
  copying nassl/__init__.py -> build/lib.linux-x86_64-2.7/nassl
  copying nassl/ssl_client.py -> build/lib.linux-x86_64-2.7/nassl
  copying nassl/debug_ssl_client.py -> build/lib.linux-x86_64-2.7/nassl
  copying nassl/x509_certificate.py -> build/lib.linux-x86_64-2.7/nassl
  copying nassl/ocsp_response.py -> build/lib.linux-x86_64-2.7/nassl
  running build_ext
  building 'nassl._nassl' extension
  creating build/temp.linux-x86_64-2.7
  creating build/temp.linux-x86_64-2.7/nassl
  creating build/temp.linux-x86_64-2.7/nassl/_nassl
  gcc -pthread -fno-strict-aliasing -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -m64 -mtune=generic -D_GNU_SOURCE -fPIC -fwrapv -DNDEBUG -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -m64 -mtune=generic -D_GNU_SOURCE -fPIC -fwrapv -fPIC -Ibin/openssl/include -Inassl/_nassl -I/usr/include/python2.7 -c nassl/_nassl/nassl.c -o build/temp.linux-x86_64-2.7/nassl/_nassl/nassl.o -Wall
  unable to execute gcc: No such file or directory
  error: command 'gcc' failed with exit status 1
  
  ----------------------------------------
  Failed building wheel for nassl
  Running setup.py clean for nassl
  Running setup.py bdist_wheel for typing ... done
  Stored in directory: /root/.cache/pip/wheels/d5/bf/c0/2219685d960a3ca52c11cbe323b2733293df7d96f97b33994e
Successfully built pshtt sslyze wget docopt typing
Failed to build nassl
Installing collected packages: requests, typing, nassl, sslyze, wget, docopt, requests-cache, pshtt
  Found existing installation: requests 2.6.0
    DEPRECATION: Uninstalling a distutils installed project (requests) has been deprecated and will be removed in a future version. This is due to the fact that uninstalling a distutils project will only partially uninstall the project.
    Uninstalling requests-2.6.0:
      Successfully uninstalled requests-2.6.0
  Running setup.py install for nassl ... error
    Complete output from command /usr/bin/python -u -c "import setuptools, tokenize;__file__='/tmp/pip-build-Izhzzl/nassl/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /tmp/pip-cKzg4G-record/install-record.txt --single-version-externally-managed --compile:
    running install
    running build
    running build_py
    creating build
    creating build/lib.linux-x86_64-2.7
    creating build/lib.linux-x86_64-2.7/nassl
    copying nassl/__init__.py -> build/lib.linux-x86_64-2.7/nassl
    copying nassl/ssl_client.py -> build/lib.linux-x86_64-2.7/nassl
    copying nassl/debug_ssl_client.py -> build/lib.linux-x86_64-2.7/nassl
    copying nassl/x509_certificate.py -> build/lib.linux-x86_64-2.7/nassl
    copying nassl/ocsp_response.py -> build/lib.linux-x86_64-2.7/nassl
    running build_ext
    building 'nassl._nassl' extension
    creating build/temp.linux-x86_64-2.7
    creating build/temp.linux-x86_64-2.7/nassl
    creating build/temp.linux-x86_64-2.7/nassl/_nassl
    gcc -pthread -fno-strict-aliasing -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -m64 -mtune=generic -D_GNU_SOURCE -fPIC -fwrapv -DNDEBUG -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -m64 -mtune=generic -D_GNU_SOURCE -fPIC -fwrapv -fPIC -Ibin/openssl/include -Inassl/_nassl -I/usr/include/python2.7 -c nassl/_nassl/nassl.c -o build/temp.linux-x86_64-2.7/nassl/_nassl/nassl.o -Wall
    unable to execute gcc: No such file or directory
    error: command 'gcc' failed with exit status 1
    
    ----------------------------------------
Command "/usr/bin/python -u -c "import setuptools, tokenize;__file__='/tmp/pip-build-Izhzzl/nassl/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /tmp/pip-cKzg4G-record/install-record.txt --single-version-externally-managed --compile" failed with error code 1 in /tmp/pip-build-Izhzzl/nassl/
[root@secengcentos7 /]#

**Comment by konklone** [comment]
> unable to execute gcc: No such file or directory

Looks like you need to install gcc. However, this isn't related to Python version support, so if you continue to have issues, please use another ticket.

**Comment by QuarantineENG** [comment]
Okay, I will have to do that after receiving the error I did on a fresh install and installing/updating all the gcc packages on the debian server.

**Comment by QuarantineENG** [comment]
I had to run - apt-get install build-essential libssl-dev libffi-dev python-dev and then run pip install pshtt to install it. 

fresh install of debian 8
apt-get update
apt-get upgrade
python -V
wget https://bootstrap.pypa.io/get-pip.py
python get-pip.py
apt-get install gcc*
apt-get upgrade
reboot -n
apt-get install build-essential libssl-dev libffi-dev python-dev
pip install pshtt

I need to get the list of domains to this server now. It worked scanning a single url.

v/r
Q

**Comment by garrettr** [comment]
@QuarantineENG Please create another issue for the issues you encountered trying to install pshtt on CentOS 7. This issue is about adding Python 3 support to pshtt, so it is not an appropriate place for you to comment about the recent issues you encountered while installing the current version of pshtt on CentOS.

**Comment by garrettr** [comment]
For what it's worth, SSLyze now has Python 3 support! See https://github.com/nabla-c0d3/sslyze/pull/203.

**Comment by konklone** [comment]
I have some beginning work on porting pshtt to Python 3 here: 

https://github.com/dhs-ncats/pshtt/compare/python3-support

I'll see if I can wrap this up soon!

**Comment by konklone** [comment]
I'm now able to get it to now fully scan a domain in Python 3, and filed #74 with a PR for review.

**Comment by h-m-f-t** [comment]
Fixed in #74.
